<a href="https://colab.research.google.com/github/PD-BDS/Deep-Learning/blob/main/DL_Assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
!pip install -q kaggle

# Data Collection

In [3]:
import pandas as pd
import os
import zipfile

from sklearn.preprocessing import OneHotEncoder, MinMaxScaler

In [ ]:
from google.colab import files

# Upload kaggle.json
files.upload()


In [5]:
!mkdir -p ~/.kaggle
!mv kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

In [6]:
!kaggle competitions download -c playground-series-s4e7 -p /content

with zipfile.ZipFile("/content/playground-series-s4e7.zip", "r") as zip_ref:
    zip_ref.extractall("/content/playground-series-s4e7")


playground-series-s4e7.zip: Skipping, found more recently modified local copy (use --force to force download)


In [7]:
df = pd.read_csv("/content/playground-series-s4e7/train.csv")
df


id  Gender  Age  Driving_License  Region_Code  \
0                0    Male   21                1         35.0   
1                1    Male   43                1         28.0   
2                2  Female   25                1         14.0   
3                3  Female   35                1          1.0   
4                4  Female   36                1         15.0   
...            ...     ...  ...              ...          ...   
11504793  11504793    Male   48                1          6.0   
11504794  11504794  Female   26                1         36.0   
11504795  11504795  Female   29                1         32.0   
11504796  11504796  Female   51                1         28.0   
11504797  11504797    Male   25                1         28.0   

          Previously_Insured Vehicle_Age Vehicle_Damage  Annual_Premium  \
0                          0    1-2 Year            Yes         65101.0   
1                          0   > 2 Years            Yes         58911.0   
2                          1    < 1 Year             No         38043.0   
3                          0    1-2 Year            Yes          2630.0   
4                          1    1-2 Year             No         31951.0   
...                      ...         ...            ...             ...   
11504793                   0    1-2 Year            Yes         27412.0   
11504794                   0    < 1 Year            Yes         29509.0   
11504795                   1    < 1 Year             No          2630.0   
11504796                   0    1-2 Year            Yes         48443.0   
11504797                   1    < 1 Year             No         32855.0   

          Policy_Sales_Channel  Vintage  Response  
0                        124.0      187         0  
1                         26.0      288         1  
2                        152.0      254         0  
3                        156.0       76         0  
4                        152.0      294         0  
...                        ...      ...       ...  
11504793                  26.0      218         0  
11504794                 152.0      115         1  
11504795                 152.0      189         0  
11504796                  26.0      274         1  
11504797                 152.0      189         0  

[11504798 rows x 12 columns]

# Data Preprocessing and Feature Engineering

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11504798 entries, 0 to 11504797
Data columns (total 12 columns):
 #   Column                Dtype  
---  ------                -----  
 0   id                    int64  
 1   Gender                object 
 2   Age                   int64  
 3   Driving_License       int64  
 4   Region_Code           float64
 5   Previously_Insured    int64  
 6   Vehicle_Age           object 
 7   Vehicle_Damage        object 
 8   Annual_Premium        float64
 9   Policy_Sales_Channel  float64
 10  Vintage               int64  
 11  Response              int64  
dtypes: float64(3), int64(6), object(3)
memory usage: 1.0+ GB


In [9]:
df.describe()

id           Age  Driving_License   Region_Code  \
count  1.150480e+07  1.150480e+07     1.150480e+07  1.150480e+07   
mean   5.752398e+06  3.838356e+01     9.980220e-01  2.641869e+01   
std    3.321149e+06  1.499346e+01     4.443120e-02  1.299159e+01   
min    0.000000e+00  2.000000e+01     0.000000e+00  0.000000e+00   
25%    2.876199e+06  2.400000e+01     1.000000e+00  1.500000e+01   
50%    5.752398e+06  3.600000e+01     1.000000e+00  2.800000e+01   
75%    8.628598e+06  4.900000e+01     1.000000e+00  3.500000e+01   
max    1.150480e+07  8.500000e+01     1.000000e+00  5.200000e+01   

       Previously_Insured  Annual_Premium  Policy_Sales_Channel       Vintage  \
count        1.150480e+07    1.150480e+07          1.150480e+07  1.150480e+07   
mean         4.629966e-01    3.046137e+04          1.124254e+02  1.638977e+02   
std          4.986289e-01    1.645475e+04          5.403571e+01  7.997953e+01   
min          0.000000e+00    2.630000e+03          1.000000e+00  1.000000e+01   
25%          0.000000e+00    2.527700e+04          2.900000e+01  9.900000e+01   
50%          0.000000e+00    3.182400e+04          1.510000e+02  1.660000e+02   
75%          1.000000e+00    3.945100e+04          1.520000e+02  2.320000e+02   
max          1.000000e+00    5.401650e+05          1.630000e+02  2.990000e+02   

           Response  
count  1.150480e+07  
mean   1.229973e-01  
std    3.284341e-01  
min    0.000000e+00  
25%    0.000000e+00  
50%    0.000000e+00  
75%    0.000000e+00  
max    1.000000e+00

In [10]:
df.isnull().sum()

id                      0
Gender                  0
Age                     0
Driving_License         0
Region_Code             0
Previously_Insured      0
Vehicle_Age             0
Vehicle_Damage          0
Annual_Premium          0
Policy_Sales_Channel    0
Vintage                 0
Response                0
dtype: int64

In [11]:
df.duplicated().sum()

0

In [12]:
df = df.drop(columns=["id"])

In [13]:
df.nunique()

Gender                      2
Age                        66
Driving_License             2
Region_Code                54
Previously_Insured          2
Vehicle_Age                 3
Vehicle_Damage              2
Annual_Premium          51728
Policy_Sales_Channel      152
Vintage                   290
Response                    2
dtype: int64

In [14]:
df["Age"].min()

20

In [15]:
df["Age"].max()

85

In [16]:
df["Age_Group"] = pd.cut(
    df["Age"],
    bins=range(20, 91, 5),
    labels=[f"{i}-{i+4}" for i in range(20, 90, 5)],
    right=False
)

# One-Hot Encoding
df = pd.get_dummies(df, columns=["Gender","Age_Group", "Vehicle_Age", "Vehicle_Damage"], drop_first=True)
df


Age  Driving_License  Region_Code  Previously_Insured  \
0          21                1         35.0                   0   
1          43                1         28.0                   0   
2          25                1         14.0                   1   
3          35                1          1.0                   0   
4          36                1         15.0                   1   
...       ...              ...          ...                 ...   
11504793   48                1          6.0                   0   
11504794   26                1         36.0                   0   
11504795   29                1         32.0                   1   
11504796   51                1         28.0                   0   
11504797   25                1         28.0                   1   

          Annual_Premium  Policy_Sales_Channel  Vintage  Response  \
0                65101.0                 124.0      187         0   
1                58911.0                  26.0      288         1   
2                38043.0                 152.0      254         0   
3                 2630.0                 156.0       76         0   
4                31951.0                 152.0      294         0   
...                  ...                   ...      ...       ...   
11504793         27412.0                  26.0      218         0   
11504794         29509.0                 152.0      115         1   
11504795          2630.0                 152.0      189         0   
11504796         48443.0                  26.0      274         1   
11504797         32855.0                 152.0      189         0   

          Gender_Male  Age_Group_25-29  ...  Age_Group_55-59  Age_Group_60-64  \
0                True            False  ...            False            False   
1                True            False  ...            False            False   
2               False             True  ...            False            False   
3               False            False  ...            False            False   
4               False            False  ...            False            False   
...               ...              ...  ...              ...              ...   
11504793         True            False  ...            False            False   
11504794        False             True  ...            False            False   
11504795        False             True  ...            False            False   
11504796        False            False  ...            False            False   
11504797         True             True  ...            False            False   

          Age_Group_65-69  Age_Group_70-74  Age_Group_75-79  Age_Group_80-84  \
0                   False            False            False            False   
1                   False            False            False            False   
2                   False            False            False            False   
3                   False            False            False            False   
4                   False            False            False            False   
...                   ...              ...              ...              ...   
11504793            False            False            False            False   
11504794            False            False            False            False   
11504795            False            False            False            False   
11504796            False            False            False            False   
11504797            False            False            False            False   

          Age_Group_85-89  Vehicle_Age_< 1 Year  Vehicle_Age_> 2 Years  \
0                   False                 False                  False   
1                   False                 False                   True   
2                   False                  True                  False   
3                   False                 False                  False   
4                   False                 False                  False  

In [17]:
num_cols = ["Age","Region_Code", "Annual_Premium", "Policy_Sales_Channel", "Vintage"]
scaler = MinMaxScaler()
df[num_cols] = scaler.fit_transform(df[num_cols])

df.head()

Age  Driving_License  Region_Code  Previously_Insured  Annual_Premium  \
0  0.015385                1     0.673077                   0        0.116218   
1  0.353846                1     0.538462                   0        0.104702   
2  0.076923                1     0.269231                   1        0.065880   
3  0.230769                1     0.019231                   0        0.000000   
4  0.246154                1     0.288462                   1        0.054547   

   Policy_Sales_Channel   Vintage  Response  Gender_Male  Age_Group_25-29  \
0              0.759259  0.612457         0         True            False   
1              0.154321  0.961938         1         True            False   
2              0.932099  0.844291         0        False             True   
3              0.956790  0.228374         0        False            False   
4              0.932099  0.982699         0        False            False   

   ...  Age_Group_55-59  Age_Group_60-64  Age_Group_65-69  Age_Group_70-74  \
0  ...            False            False            False            False   
1  ...            False            False            False            False   
2  ...            False            False            False            False   
3  ...            False            False            False            False   
4  ...            False            False            False            False   

   Age_Group_75-79  Age_Group_80-84  Age_Group_85-89  Vehicle_Age_< 1 Year  \
0            False            False            False                 False   
1            False            False            False                 False   
2            False            False            False                  True   
3            False            False            False                 False   
4            False            False            False                 False   

   Vehicle_Age_> 2 Years  Vehicle_Damage_Yes  
0                  False                True  
1                   True                True  
2                  False               False  
3                  False                True  
4                  False               False  

[5 rows x 25 columns]

In [18]:
df['Response'].value_counts()

Response
0    10089739
1     1415059
Name: count, dtype: int64

In [19]:
from sklearn.model_selection import train_test_split

# Sampling
sample_size = 500000  # Taking 500k rows from 11.5M

df_sample, _ = train_test_split(df, train_size=sample_size, stratify=df["Response"], random_state=42)

print(f"Sampled Dataset Shape: {df_sample.shape}")


Sampled Dataset Shape: (500000, 25)


# Train - Validation - Test Split

In [20]:
x = df_sample.drop(columns=["Response"])
y = df_sample["Response"]

# First, split into Train (80%) and Test (20%)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=40, stratify=y)

# Then, split the train set into Train (80%) and Validation (20%)
x_train, x_val, y_train, y_val = train_test_split(x_train, y_train, test_size=0.2, random_state=40, stratify=y_train)

print(f"Training Set: {x_train.shape}, Validation Set: {x_val.shape}, Test Set: {x_test.shape}")


Training Set: (320000, 24), Validation Set: (80000, 24), Test Set: (100000, 24)


# MPL Modeling

In [21]:
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.utils.data import DataLoader, TensorDataset, random_split
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report
from sklearn.utils.class_weight import compute_class_weight
from imblearn.over_sampling import ADASYN
from imblearn.under_sampling import EditedNearestNeighbours

In [22]:
# Applying ADASYN to generate synthetic samples
adasyn = ADASYN(sampling_strategy='minority', random_state=40)
x_train_adasyn, y_train_adasyn = adasyn.fit_resample(x_train, y_train)

In [23]:
scaler = StandardScaler()

# Fit on training data & transform
x_train_scaled = scaler.fit_transform(x_train_adasyn)
x_val_scaled = scaler.transform(x_val)
x_test_scaled = scaler.transform(x_test)

In [24]:
y_train = y_train.values
y_val = y_val.values
y_test = y_test.values



In [25]:
# Convert data to PyTorch tensors
x_train_tensor = torch.tensor(x_train_scaled, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train_adasyn, dtype=torch.long)

x_val_tensor = torch.tensor(x_val_scaled, dtype=torch.float32)
y_val_tensor = torch.tensor(y_val, dtype=torch.long)

x_test_tensor = torch.tensor(x_test_scaled, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test, dtype=torch.long)

# Create PyTorch datasets
train_dataset = TensorDataset(x_train_tensor, y_train_tensor)
val_dataset = TensorDataset(x_val_tensor, y_val_tensor)
test_dataset = TensorDataset(x_test_tensor, y_test_tensor)

# Define DataLoaders
batch_size = 128
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)


In [26]:
import torch.nn as nn
import torch.nn.functional as F

class MLP(nn.Module):
    def __init__(self, input_size):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(input_size, 128)
        self.bn1 = nn.BatchNorm1d(128)
        self.fc2 = nn.Linear(128, 64)
        self.bn2 = nn.BatchNorm1d(64)
        self.fc3 = nn.Linear(64, 32)
        self.fc4 = nn.Linear(32, 2)  # Binary classification (0 or 1)

        self.dropout = nn.Dropout(0.2)

    def forward(self, x):
        x = F.relu(self.bn1(self.fc1(x)))
        x = self.dropout(x)
        x = F.relu(self.bn2(self.fc2(x)))
        x = self.dropout(x)
        x = F.relu(self.fc3(x))
        x = self.fc4(x)  # Output layer
        return x


In [27]:
import torch.optim as optim

def train_model(model, train_loader, val_loader, criterion, optimizer, epochs=10):
    train_losses, val_losses = [], []

    for epoch in range(epochs):
        model.train()
        train_loss = 0.0

        for x_batch, y_batch in train_loader:
            optimizer.zero_grad()
            outputs = model(x_batch)
            loss = criterion(outputs, y_batch)
            loss.backward()
            optimizer.step()
            train_loss += loss.item()

        train_loss /= len(train_loader)
        train_losses.append(train_loss)

        # Validation Loss
        model.eval()
        val_loss = 0.0
        with torch.no_grad():
            for x_batch, y_batch in val_loader:
                outputs = model(x_batch)
                loss = criterion(outputs, y_batch)
                val_loss += loss.item()

        val_loss /= len(val_loader)
        val_losses.append(val_loss)

        print(f"Epoch [{epoch+1}/{epochs}] - Train Loss: {train_loss:.4f}, Val Loss: {val_loss:.4f}")

    return train_losses, val_losses


In [28]:
# Initialize Model
input_size = x_train_adasyn.shape[1]  # Number of features
model = MLP(input_size)

# Loss Function & Optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Train Model
train_losses, val_losses = train_model(model, train_loader, val_loader, criterion, optimizer, epochs=10)


Epoch [1/10] - Train Loss: 0.4311, Val Loss: 0.4794
Epoch [2/10] - Train Loss: 0.4230, Val Loss: 0.4362
Epoch [3/10] - Train Loss: 0.4178, Val Loss: 0.4619
Epoch [4/10] - Train Loss: 0.4154, Val Loss: 0.4451
Epoch [5/10] - Train Loss: 0.4139, Val Loss: 0.4660
Epoch [6/10] - Train Loss: 0.4128, Val Loss: 0.4605
Epoch [7/10] - Train Loss: 0.4114, Val Loss: 0.4516
Epoch [8/10] - Train Loss: 0.4105, Val Loss: 0.4546
Epoch [9/10] - Train Loss: 0.4091, Val Loss: 0.4728
Epoch [10/10] - Train Loss: 0.4078, Val Loss: 0.4427


In [29]:
from sklearn.metrics import accuracy_score, classification_report

def evaluate_model(model, test_loader):
    model.eval()
    y_pred, y_true = [], []

    with torch.no_grad():
        for x_batch, y_batch in test_loader:
            outputs = model(x_batch)
            preds = torch.argmax(outputs, dim=1)
            y_pred.extend(preds.tolist())
            y_true.extend(y_batch.tolist())

    accuracy = accuracy_score(y_true, y_pred)
    print(f"Test Accuracy: {accuracy:.4f}")
    print(classification_report(y_true, y_pred))

evaluate_model(model, test_loader)


Test Accuracy: 0.6984
              precision    recall  f1-score   support

           0       0.98      0.67      0.80     87700
           1       0.28      0.92      0.43     12300

    accuracy                           0.70    100000
   macro avg       0.63      0.80      0.61    100000
weighted avg       0.90      0.70      0.75    100000



In [30]:
from sklearn.metrics import precision_recall_curve, auc

# Get the predicted probabilities for the positive class
y_pred_proba = model(x_test_tensor).softmax(dim=1)[:, 1].detach().numpy()

# Calculate precision-recall curve
precision, recall, _ = precision_recall_curve(y_test_tensor.numpy(), y_pred_proba)

# Calculate AUC score
pr_auc = auc(recall, precision)

print(f'Precision-Recall AUC: {pr_auc:.4f}')


Precision-Recall AUC: 0.3480


# Hyperparameter Experimentation

In [31]:
df = df_sample

In [32]:
import torch.optim as optim
import torch.nn.functional as F
import torch.nn as nn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.utils.class_weight import compute_class_weight
from imblearn.over_sampling import ADASYN
import numpy as np
from torch.utils.data import DataLoader, TensorDataset
from sklearn.metrics import classification_report

# Define hyperparameters to experiment with
batch_sizes = [64, 128, 256]
learning_rates = [0.001, 0.0005, 0.01]
epochs = 10

# Early stopping patience (number of epochs to wait for improvement before stopping)
early_stopping_patience = 5

# Function to experiment with hyperparameters
def experiment_model(batch_size, lr):
    # Recreate train/test splits
    x = df.drop(columns=["Response"])
    y = df["Response"]

    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=40, stratify=y)
    x_train, x_val, y_train, y_val = train_test_split(x_train, y_train, test_size=0.2, random_state=40, stratify=y_train)

    # Scaling
    scaler = StandardScaler()
    x_train_scaled = scaler.fit_transform(x_train)
    x_val_scaled = scaler.transform(x_val)
    x_test_scaled = scaler.transform(x_test)

    # ADASYN oversampling
    adasyn = ADASYN(sampling_strategy='minority', random_state=40)
    x_train_adasyn, y_train_adasyn = adasyn.fit_resample(x_train_scaled, y_train)

    # Convert data to tensors
    x_train_tensor = torch.tensor(x_train_adasyn, dtype=torch.float32)
    y_train_tensor = torch.tensor(y_train_adasyn, dtype=torch.long)
    x_val_tensor = torch.tensor(x_val_scaled, dtype=torch.float32)
    y_val_tensor = torch.tensor(y_val.values, dtype=torch.long)
    x_test_tensor = torch.tensor(x_test_scaled, dtype=torch.float32)
    y_test_tensor = torch.tensor(y_test.values, dtype=torch.long)

    # Datasets and DataLoaders
    train_dataset = TensorDataset(x_train_tensor, y_train_tensor)
    val_dataset = TensorDataset(x_val_tensor, y_val_tensor)
    test_dataset = TensorDataset(x_test_tensor, y_test_tensor)

    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

    # Define the model architecture
    class MLP(nn.Module):
        def __init__(self, input_size, hidden_layers=[128, 64, 32]):
            super(MLP, self).__init__()
            self.fc1 = nn.Linear(input_size, hidden_layers[0])
            self.fc2 = nn.Linear(hidden_layers[0], hidden_layers[1])
            self.fc3 = nn.Linear(hidden_layers[1], hidden_layers[2])
            self.fc4 = nn.Linear(hidden_layers[2], 2)  # Binary classification (0 or 1)
            self.dropout = nn.Dropout(0.3)
            self.bn1 = nn.BatchNorm1d(hidden_layers[0])
            self.bn2 = nn.BatchNorm1d(hidden_layers[1])
            self.bn3 = nn.BatchNorm1d(hidden_layers[2])

        def forward(self, x):
            x = F.relu(self.bn1(self.fc1(x)))
            x = self.dropout(x)
            x = F.relu(self.bn2(self.fc2(x)))
            x = self.dropout(x)
            x = F.relu(self.bn3(self.fc3(x)))
            x = self.fc4(x)  # Output layer
            return x

    # Initialize Model
    model = MLP(input_size=x_train.shape[1])

    # Loss Function & Optimizer
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=lr)

    # Early stopping mechanism
    best_val_loss = float('inf')
    patience_counter = 0

    # Train Model
    train_losses, val_losses = [], []
    for epoch in range(epochs):
        model.train()
        train_loss = 0.0

        for x_batch, y_batch in train_loader:
            optimizer.zero_grad()
            outputs = model(x_batch)
            loss = criterion(outputs, y_batch)
            loss.backward()
            optimizer.step()
            train_loss += loss.item()

        train_loss /= len(train_loader)
        train_losses.append(train_loss)

        # Validation Loss
        model.eval()
        val_loss = 0.0
        with torch.no_grad():
            for x_batch, y_batch in val_loader:
                outputs = model(x_batch)
                loss = criterion(outputs, y_batch)
                val_loss += loss.item()

        val_loss /= len(val_loader)
        val_losses.append(val_loss)

        print(f"Epoch [{epoch+1}/{epochs}] - Train Loss: {train_loss:.4f}, Val Loss: {val_loss:.4f}")

        # Early stopping check
        if val_loss < best_val_loss:
            best_val_loss = val_loss
            patience_counter = 0
        else:
            patience_counter += 1
            if patience_counter >= early_stopping_patience:
                print("Early stopping due to no improvement in validation loss.")
                break

    # Test the model
    model.eval()
    correct, total = 0, 0
    with torch.no_grad():
        for x_batch, y_batch in test_loader:
            outputs = model(x_batch)
            _, predicted = torch.max(outputs, 1)
            total += y_batch.size(0)
            correct += (predicted == y_batch).sum().item()

    accuracy = correct / total
    print(f"Test Accuracy: {accuracy:.4f}")

    # Classification report
    y_true = y_test_tensor.numpy()
    y_pred = []
    with torch.no_grad():
        for x_batch in test_loader:
            outputs = model(x_batch[0])
            _, predicted = torch.max(outputs, 1)
            y_pred.extend(predicted.numpy())

    print(classification_report(y_true, y_pred))

# Experiment with different batch sizes and learning rates
for batch_size in batch_sizes:
    for lr in learning_rates:
        print(f"Running experiment with batch_size={batch_size} and learning_rate={lr}")
        experiment_model(batch_size, lr)


Running experiment with batch_size=64 and learning_rate=0.001
Epoch [1/10] - Train Loss: 0.4372, Val Loss: 0.4888
Epoch [2/10] - Train Loss: 0.4323, Val Loss: 0.4682
Epoch [3/10] - Train Loss: 0.4307, Val Loss: 0.4598
Epoch [4/10] - Train Loss: 0.4299, Val Loss: 0.4668
Epoch [5/10] - Train Loss: 0.4290, Val Loss: 0.4731
Epoch [6/10] - Train Loss: 0.4283, Val Loss: 0.4636
Epoch [7/10] - Train Loss: 0.4279, Val Loss: 0.4724
Epoch [8/10] - Train Loss: 0.4274, Val Loss: 0.4776
Early stopping due to no improvement in validation loss.
Test Accuracy: 0.6727
              precision    recall  f1-score   support

           0       0.99      0.63      0.77     87700
           1       0.27      0.96      0.42     12300

    accuracy                           0.67    100000
   macro avg       0.63      0.80      0.60    100000
weighted avg       0.90      0.67      0.73    100000

Running experiment with batch_size=64 and learning_rate=0.0005
Epoch [1/10] - Train Loss: 0.4388, Val Loss: 0.4630
E



*   The Dataset is highly imbalanced, therefore ADASYN was implemented to synthetically represent minority class. Before implementing adasyn the model was resulting 85% accuracy, but it was unable to generalize and accurately predict the minority class.
*   After implementing ADASYN, the accuracy dropped to 68-70%, using multiple hyperparameter tuning, even though the overall accuracy dropped, the model's ability to predict the minority class has improved.
*   however, There is still room for improvement to optimize the model. trying more hidden layer combination and using other activation function such as sigmoid, tanh etc could improve the model's performance.
*   To, conclude, follwing the standard procedure to train a model using pytorch, a decent model has been trained, but its hasn't been able to predict optimum results yet.

